drivers/usb/usbstor_new/error.c

```diff
@@ -107,304 +60,209 @@ USBSTOR_ResetPipeWithHandle(
-        Context->Irp->IoStatus.Information = 0;
-        USBSTOR_QueueTerminateRequest(Context->PDODeviceExtension->LowerDeviceObject, Context->Irp);
-        IoCompleteRequest(Context->Irp, IO_NO_INCREMENT);
+    //FIXME RemoveLock
+}
 
-        //
-        // Start the next request
-        //
-        USBSTOR_QueueNextRequest(Context->PDODeviceExtension->LowerDeviceObject);
+VOID
+NTAPI
+USBSTOR_BulkQueueResetPipe(
+    IN PFDO_DEVICE_EXTENSION FDODeviceExtension)
+{
+    DPRINT("USBSTOR_BulkQueueResetPipe: \n");
 
-        //
-        // srb handling finished
-        //
-        Context->FDODeviceExtension->SrbErrorHandlingActive = FALSE;
+    //FIXME RemoveLock
 
-        //
-        // clear timer srb
-        //
-        Context->FDODeviceExtension->LastTimerActiveSrb = NULL;
-    }
-    else
+    IoQueueWorkItem(FDODeviceExtension->ResetDeviceWorkItem,
+                    USBSTOR_BulkResetPipeWorkItem,
+                    CriticalWorkQueue,
+                    0);
+}
+
+VOID
+NTAPI
+USBSTOR_ResetDeviceWorkItem(
+    IN PDEVICE_OBJECT FdoDevice,
+    IN PVOID Context)
+{
+    PFDO_DEVICE_EXTENSION FDODeviceExtension;
+    PPDO_DEVICE_EXTENSION PDODeviceExtension;
+    ULONG nx;
+    NTSTATUS Status;
+    KIRQL OldIrql;
+
+    DPRINT("USBSTOR_ResetDeviceWorkItem: \n");
+
+    PDODeviceExtension = (PPDO_DEVICE_EXTENSION)Context;
+    FDODeviceExtension = FdoDevice->DeviceExtension;
+
+    if (FDODeviceExtension->CurrentIrp)
     {
-        DPRINT1("Retrying Count %lu %p\n", Context->RetryCount, Context->PDODeviceExtension->Self);
+        IoCancelIrp(FDODeviceExtension->CurrentIrp);
 
-        //
-        // re-schedule request
-        //
-        USBSTOR_HandleExecuteSCSI(Context->PDODeviceExtension->Self, Context->Irp, Context->RetryCount + 1);
+        KeWaitForSingleObject(&FDODeviceExtension->TimeOutEvent,
+                              Executive,
+                              KernelMode,
+                              FALSE,
+                              NULL);
 
-        //
-        // srb error handling finished
-        //
-        Context->FDODeviceExtension->SrbErrorHandlingActive = FALSE;
+        if (!Context)
+        {
+            PIO_STACK_LOCATION IoStack;
 
-        //
-        // srb error handling finished
-        //
-        Context->FDODeviceExtension->TimerWorkQueueEnabled = TRUE;
+            IoStack = IoGetCurrentIrpStackLocation(FDODeviceExtension->CurrentIrp);
+            PDODeviceExtension = IoStack->Parameters.Others.Argument2;
+        }
 
-        //
-        // clear timer srb
-        //
-        Context->FDODeviceExtension->LastTimerActiveSrb = NULL;
+        KeRaiseIrql(DISPATCH_LEVEL, &OldIrql);
+        IoCompleteRequest(FDODeviceExtension->CurrentIrp, IO_NO_INCREMENT);
+        KeLowerIrql(OldIrql);
+
+        FDODeviceExtension->CurrentIrp = NULL;
     }
 
-    //
-    // cleanup irp context
-    //
-    FreeItem(Context->cbw);
-    FreeItem(Context);
+    nx = 0;
 
+    do
```

Comment:
Together with the previous "nx = 0;" allocation, this loop can be rewritten as:
```
for (nx = 0; nx < 3; ++nx)
{
    // Code of the do-while loop
}
```

Reply:
Rewritted